### PR TITLE
Make trigger service auth middleware /login endpoint compatible with Auth0

### PR DESCRIPTION
**Pull request by aherrmann-da**
- Make the authorization and token endpoints configurable
- Use `application/x-www-form-urlencoded` as specified in https://tools.ietf.org/html/rfc6749#section-4.1.3
- Check the status code of the token endpoint response
- Fix the type of the token's `expires_in` field (`Int` instead of `String`)
- Adds instructions for manual testing against Auth0

Part of https://github.com/digital-asset/daml/issues/6923

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala [hunk 1]
```diff
@@ -119,13 +116,22 @@ object Server extends StrictLogging {
...
               val tokenRequest = for {
+                entity <- Marshal(body).to[RequestEntity]
+                req = HttpRequest(
+                  uri = config.oauthToken,
+                  entity = entity,
+                  method = HttpMethods.POST)
                 resp <- Http().singleRequest(req)
-                tokenResp <- Unmarshal(resp).to[OAuthResponse.Token]
+                tokenResp <- if (resp.status != StatusCodes.OK) {
+                  Unmarshal(resp).to[String].flatMap { msg =>
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:128]: I still feel like there must be some builtin pattern for checking the status code and then unmarshalling but I never found one :confused:

File: triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala [hunk 2]
```diff
@@ -60,7 +85,7 @@ object Response {
   case class Token(
       accessToken: String,
       tokenType: String,
-      expiresIn: Option[String],
+      expiresIn: Option[Int],
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala:88]: good catch!

**Review by cocreature — APPROVED** [review]
Great to see this already working, thanks!

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+repository](https://github.com/digital-asset/ex-secure-daml-infra).
+
+### Setup
+
+* Sign up for an account on [Auth0](https://auth0.com).
+* Create a new API.
+  - Provide a name (`ex-daml-api`).
+  - Provide an Identifier (`https://daml.com/ledger-api`).
+  - Select Signing Algorithm of `RS256`.
+* Create a new machine to machine application.
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md:21]: "Machine to Machine" is one of the available application types that Auth0 provides. It looks like "Native" is closer to what we want here. At least it enables application code grant by default.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    @param {object} context - additional authorization context
+    @param {object} context.webtask - webtask context
+    @param {function} cb - function (error, accessTokenClaims)
+    */
+    module.exports = function(client, scope, audience, context, cb) {
+      var access_token = {};
+      var actAs = [];
+      var readAs = [];
+      var admin = false;
+      // TODO define mapping from scope to ledger claims.
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md:51]: Right, the implementation is just the first naive mapping that came to mind. The todo note is about putting some thought into what this mapping should look like. I'm planning to do that when integrating with the trigger service.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      --oauth-auth AUTHURL \
+      --oauth-token TOKENURL
+  ```
+  - Replace `CLIENTID` and `CLIENTSECRET` by the "Client ID" and "Client
+    Secret" from above.
+  - Replace `AUTHURL` and `TOKENURL` by the "OAuth Authorization URL"
+    and "OAuth Token URL" from above. They should look as follows:
+    ```
+    https://XYZ.auth0.com/authorize
+    https://XYZ.auth0.com/oauth/token
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md:94]: Indeed, according to [this SO question](https://stackoverflow.com/questions/40797181/google-oauth-2-0-new-authorization-and-token-endpoint/40800830) you can even find instances where the authority differs between them.

**Pull request merged**